File: packages/core/src/pomi/injector.ts
import type { Canary, ChallengePayload } from '../types.js'
import { CanaryCatalog, type CatalogSelectOptions } from './catalog.js'

export interface InjectionResult {
  payload: ChallengePayload
  injected: Canary[]
}

export interface InjectOptions {
  exclude?: string[]
}

export class CanaryInjector {
  private catalog: CanaryCatalog

  constructor(catalog: CanaryCatalog) {
    this.catalog = catalog
  }

  inject(payload: ChallengePayload, count: number, options?: InjectOptions): InjectionResult {
    if (count <= 0) {
      return { payload: { ...payload }, injected: [] }
    }

    const selected = this.catalog.select(count, { exclude: options?.exclude })
    if (selected.length === 0) {
      return { payload: { ...payload }, injected: [] }
    }

    // Group canaries by injection method
    const prefixCanaries = selected.filter((c) => c.injection_method === 'prefix')
    const inlineCanaries = selected.filter((c) => c.injection_method === 'inline')
    const suffixCanaries = selected.filter((c) => c.injection_method === 'suffix')
    const embeddedCanaries = selected.filter((c) => c.injection_method === 'embedded')

    let instructions = payload.instructions

    // Prefix: add before main instructions
    if (prefixCanaries.length > 0) {
      const prefixText = prefixCanaries
        .map((c) => `- ${c.id}: ${c.prompt}`)
        .join('\n')
      instructions = `Before starting, answer these briefly (include in canary_responses):\n${prefixText}\n\n${instructions}`
    }

    // Inline & Suffix & Embedded: add as "Side tasks" after main instructions
    const sideTaskCanaries = [...inlineCanaries, ...suffixCanaries, ...embeddedCanaries]
    if (sideTaskCanaries.length > 0) {
      const sideText = sideTaskCanaries
        .map((c) => `- ${c.id}: ${c.prompt}`)
        .join('\n')
      instructions = `${instructions}\n\nAlso, complete these side tasks (include answers in canary_responses field):\n${sideText}`
    }

    const newPayload: ChallengePayload = {
      ...payload,
      instructions,
      context: {
        ...payload.context,
        canary_ids: selected.map((c) => c.id),
      },
    }

    return { payload: newPayload, injected: selected }
  }
}

File: sdks/python/src/xagentauth/pomi/injector.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from xagentauth.pomi.catalog import CanaryCatalog, CatalogSelectOptions
from xagentauth.types import Canary, ChallengePayload


@dataclass
class InjectionResult:
    payload: ChallengePayload
    injected: list[Canary]


class CanaryInjector:
    """Injects canary probes into challenge payloads."""

    def __init__(self, catalog: CanaryCatalog) -> None:
        self._catalog = catalog

    def inject(
        self,
        payload: ChallengePayload,
        count: int,
        exclude: Optional[list[str]] = None,
    ) -> InjectionResult:
        if count <= 0:
            return InjectionResult(
                payload=payload.model_copy(),
                injected=[],
            )

        options = CatalogSelectOptions(exclude=exclude) if exclude else None
        selected = self._catalog.select(count, options)
        if len(selected) == 0:
            return InjectionResult(
                payload=payload.model_copy(),
                injected=[],
            )

        # Group canaries by injection method
        prefix_canaries = [c for c in selected if c.injection_method == "prefix"]
        inline_canaries = [c for c in selected if c.injection_method == "inline"]
        suffix_canaries = [c for c in selected if c.injection_method == "suffix"]
        embedded_canaries = [c for c in selected if c.injection_method == "embedded"]

        instructions = payload.instructions

        # Prefix: add before main instructions
        if prefix_canaries:
            prefix_text = "\n".join(f"- {c.id}: {c.prompt}" for c in prefix_canaries)
            instructions = (
                f"Before starting, answer these briefly (include in canary_responses):\n"
                f"{prefix_text}\n\n{instructions}"
            )

        # Inline & Suffix & Embedded: add as "Side tasks" after main instructions
        side_task_canaries = inline_canaries + suffix_canaries + embedded_canaries
        if side_task_canaries:
            side_text = "\n".join(f"- {c.id}: {c.prompt}" for c in side_task_canaries)
            instructions = (
                f"{instructions}\n\n"
                f"Also, complete these side tasks (include answers in canary_responses field):\n"
                f"{side_text}"
            )

        new_context = dict(payload.context or {})
        new_context["canary_ids"] = [c.id for c in selected]

        new_payload = ChallengePayload(
            type=payload.type,
            instructions=instructions,
            data=payload.data,
            steps=payload.steps,
            context=new_context,
        )

        return InjectionResult(payload=new_payload, injected=selected)
